service_discovery/service_import/service_import_utils.go

```diff
@@ -0,0 +1,303 @@
+}
+
+func (csic *ClusterServiceImportCache) DeleteServiceImport(cname, ns, name string) {
+	if sic := csic.GetServiceImportCache(cname); sic != nil {
+		csic.lock.RLock()
+		defer csic.lock.RUnlock()
+		sic.DeleteServiceImport(ns, name)
+		return
+	}
+}
+
+func (csic *ClusterServiceImportCache) GetServiceImport(cname, ns, name string) *siapi.ServiceImport {
+	if sic := csic.GetServiceImportCache(cname); sic != nil {
+		return sic.LookupServiceImport(ns, name)
+	}
+	return nil
+}
+
+type ServiceImportHandler struct {
+	clusterServiceImportCache *ClusterServiceImportCache
+	amkoClientset             *amkov1.Clientset
+	serviceImportController   *ServiceImportController
+}
+
+var serviceImportHandler *ServiceImportHandler
+var sihOnce sync.Once
+
+func InitServiceImportHandler(amkoClientset *amkov1.Clientset, clusterList []string,
+	siCtrl *ServiceImportController) {
+	sihOnce.Do(func() {
+		serviceImportHandler = &ServiceImportHandler{
+			clusterServiceImportCache: InitClusterServiceImportCache(clusterList),
+			amkoClientset:             amkoClientset,
+			serviceImportController:   siCtrl,
+		}
+	})
+}
+
+func GetServiceImportHandler() *ServiceImportHandler {
+	return serviceImportHandler
+}
+
+// If we encounter a new service from a cluster, we need to check the corresponding service import
+// object at a couple of places: the service import cache and the informer cache.
+//
+// If the service import object exists in the service import cache but not in the informer cache,
+// we need to first update the service import cache and then create a new service import object
+// in the cluster.
+// If the service import object doesn't exist in the service import cache, but exists in the
+// informer cache, we need to add this object to the service import cache and check if an update
+// is required for the corresponding object in the informer cache.
+// If the service import object exists at both places, we need to check if an endpoint update
+// is required in the service import cache. If yes, update the endpoints in the service import
+// cache and then in the cluster's service import object.
+// If the service import doesn't exist in either place, create a new one in both places.
+func (sih *ServiceImportHandler) AddUpdateService(obj *siapi.ServiceImport) error {
+	oldObj, err := sih.serviceImportController.GetServiceImportObjectFromInformerCache(obj.Spec.Cluster,
+		obj.Spec.Namespace, obj.Spec.Service)
+	if err != nil {
+		if k8serrors.IsNotFound(err) {
+			// object not found, will create one
+			err := sih.serviceImportController.CreateServiceImportObj(obj)
+			if err != nil {
+				return err
+			}
+			// add this object to the cache
+			sih.clusterServiceImportCache.AddUpdateServiceImport(obj)
+			gslbutils.Logf("ns: %s, name: %s, msg: service import object created successfully", obj.GetNamespace(),
+				obj.GetName())
+			return nil
+		}
+		return fmt.Errorf("error in getting object from informer cache: %v", err)
```

Comment:
Do we need to put an else block here ? Not found is not an error which would need a retry